### DISCUSSION TITLE
A new wave of analyzers in .NET 8

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#29] [General]
# A new wave of analyzers in .NET 8

Analyzers did become an integral part of the .NET ecosystem. Their main responsibility is to find potential code issues and warn you. Often times this comes even with potential fixes you can directly apply.
And Microsoft will continue this journey with the upcoming .NET 8 release. This blog post will show you potential candidates, which will make the cut.

https://steven-giesel.com/blogPost/1a03a998-e428-4a42-9c30-1cfbf6ac5f4c

**Comment by ndepend** [comment]
Concerning the ToLower() stuff, there is also a trap with special characters like the Turkish I we stumbled upon once with a Turkish  user
https://haacked.com/archive/2012/07/05/turkish-i-problem-and-why-you-should-care.aspx/

**Reply by linkdotnet** [reply]
Yes that's true.`ToLowerInvariant` and its brother `ToUpperInvariant` don't suffer from the Turkish "i" problem, but still would be flagged by the compiler as you still create a new string.

**Comment by darthwalsh** [comment]
The Any() rule feels like a micro-optimization. If I was considering that, I might also consider migrating that performance-critical code to C++ or Rust. 
(BTW, I think the GitHub link for that rule isn't correct.)

**Reply by linkdotnet** [reply]
First of all thanks for reporting the broken link. It is fixed now. 

The given severity is "Info", but I am not sure if this is maybe too strong. I am on your side @darthwalsh, the level of optimization is far beyond your normal levels. I do believe `Any` is more concise and oftentimes more readable (in my opinion).

Stephen Toub argues, that `Any()`, in contrast to `.Length != 0` is less readable and can hide the intent.

**Reply by JustDre** [reply]
Definitely .Any() is more clear.  That it checks the type and cannot be inlined to .Length != 0 is perhaps a result of the type system, and won't be obvious without inspecting the compiled and JITed code, possibly even requiring an understanding of why Roslyn generates type checks.  I use .Any() with the understanding that the code *should* compile to the same thing, if not now then in the future.  That Stephen Toub better understands the actual implementation of .Any() is probably the reason he argues that .Length != 0 is more clear, as it's more explicit.  Perhaps we need an extension method that more succinctly expresses the intent, and can still be inlined to the expected.

**Reply by linkdotnet** [reply]
My best guess why `Any` is not inline-able, is because of this:
```csharp
using (IEnumerator<TSource> e = source.GetEnumerator()) 
```

Taken from [here](https://github.com/microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs#L1288). `using` is basically a try-finally block. Any try block can't be inlined. 

Anyway, I also **never** saw that `Any` would lead to such side-effects as described (where I am also not 100% sure, if I really understand the statement). As we are still in the alpha-phase of .NET 8, everything can change.

**Reply by kaspervandenberg** [reply]
Would it be possible to short circuit `Any()` with `Length != 0` (or `Count != 0`) for (commonly used) types of enumerable that support it?  (Ofcourse without checking the type of container at runtime.  Currently I don't know if extension methods support overloading.)

**Reply by linkdotnet** [reply]
`Any` is an extension method so that should work...given the fact that there are already overloaded versions of `Any`. Good point @kaspervandenberg

**Reply by JustDre** [reply]
.NET 6 added TryGetNonEnumeratedCount, which is used in .Any().  https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/AnyAll.cs

But as mentioned, the entire method cannot be inlined, either because of the *using* or else because it's too long.  I've verified in SharpLab that an extension method such as the following does indeed get inlined:

```public static bool HasAny<T>(this List<T> list) => list.Count != 0; ```

**Reply by linkdotnet** [reply]
Hmm good to know and thanks for the analysis @JustDre

I am inclined to do so some performance tests/benchmarks in the upcoming days. I am just interested in what this analyzer suggestion means in numbers. Anyway, it is only flagged as "info" as of now.

**Comment by Caltor** [comment]
I realise it's not the point of the article but I think the logic for these two lines is inverted:

```
var isEmpty = names.Any();
var isEmpty = names.Count != 0;
```

I think it should be:

```
var isEmpty = !names.Any();
var isEmpty = names.Count == 0;
```

Or rename `isEmpty` to something more suitable.

**Reply by linkdotnet** [reply]
Hey @Caltor , you are absolutely right! Spot on and thanks for reporting! Fixed

**Comment by DrGriff** [comment]
Regarding `.Any()` vs `.Count != 0`.  If my collection contained 1 million items, then would it not need to iterate the whole collection to calculate .Count, whereas .Any() would stop executing as soon as it found the first item?

**Reply by linkdotnet** [reply]
It really depends on your underlying collection. Let's say you have a List or Array then `Count` is just a property that is "constantly updated" when you add or delete items. So no one has to go through your items to check, as this information is available in constant time (O(1) in big O notation).

`Any` has also some shortcuts if it detects that the underlying type is a List, Array (and everything that is materialized). If so, it just checks `Count != 0`. 

The picture shifts if you have something like `IEnumerable<T>`. Here `Any()` just has to check one single entry to return true or false. But you might notice that `IEnumerable` doesn't offer a `Count` or `Length` property - as this information would require to go through the whole enumeration. If the underlying collection is a database, that might be a huge impact to your application.

That is why the given analyzer is only checking for types that are already materialized (`List`, `Array` and friends).

I hope that answers your question.

**Comment by fseidl-bauradar** [comment]
A little bit of the toppic, but I waiting for a new solution how to retrieve the HttpContext over the DI outside the Controllers after the IHttpContextAccessor is no longer available

**Reply by linkdotnet** [reply]
Can you a give a bit more context about your usecase? I'd like to understand your requirements a bit better.

**Reply by fseidl-bauradar** [reply]
I require some Data given inside the Http Parameters of the Request, but I don't want to access it directly because I want to use the Context directly, therefore I previously registered the IHttpContextAccessor inside the DI to request the context inside a helper class, and extract the data there and provide it to the Service Layer Classes.

**Reply by fseidl-bauradar** [reply]
Now after the HttpContextAccessor has been marked as Deprecated, I am looking for a new solution to replace the usage.

**Reply by linkdotnet** [reply]
Maybe I missed something but the `IHttpContextAccessor` ins't deprecated, is it?

The Nuget package that provided it is deprecated, true.
But you can still add this to your csproj:

```xml
<ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
</ItemGroup>
```

Which gives you all the objects provided by the shared ASP.NET Framework.

**Reply by fseidl-bauradar** [reply]
![grafik](https://github.com/linkdotnet/Blog.Discussions/assets/125438559/6a46cf39-0673-42cb-93ca-3107bee7fcd6)

The package is marked as deprecated and there is no new Package replacing it.

You can still install it, but without bug fixing, it is not adviseable to use it.

**Reply by linkdotnet** [reply]
Exactly the NuGet package is deprecated/obsolete.
The way forward is opting in into the shared framework as described above (via `FrameworkRefernece`) .
With that you always have the same version as the target framework.

**Reply by fseidl-bauradar** [reply]
You mean the IHttpContextHelper is also included inside the Microsoft.AspNetCore.App package, nice too know

**Reply by fseidl-bauradar** [reply]
![grafik](https://github.com/linkdotnet/Blog.Discussions/assets/125438559/46b1e5b5-e0d1-4dec-88d6-15743dd46754)

But there it's even worse, it's deprecated too already and a lot Security holes are reported by CheckMarks

**Reply by linkdotnet** [reply]
If you mean `IHttpContextAccessor` - then yes.

If you create a new ASP.NET Web Project, you can freely use the `IHttpContextAccessor`. 
Basically the:
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
```

Is a convenient shorthand for:
```xml
<ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
</ItemGroup>
```
Plus a bit more.

**Reply by linkdotnet** [reply]
Nono - not the NuGet package. That has nothing to do with any external dependency.

The package you are referring was a "polyfill" prior to .net core 3.
This:
```xml
<ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
</ItemGroup>
```

Has nothing to do with NuGet directly. Depending on your Target Framework, you reference those assemblies implicitly.

So to sum it up, there are two ways to access your `IHttpContextAccessor` without **any** Nuget package:

1. Use the `<Project Sdk="Microsoft.NET.Sdk.Web">` SDK type - aka you create a web application. Then you can easily call: `builder.Services.AddHttpContextAccessor()` and use the interface.
2. Or if you are in a library project, use the `<FrameworkReference Include="Microsoft.AspNetCore.App" />` part.

**Reply by linkdotnet** [reply]
When you install the .NET SDK locally you are also installing assemblies for ASP.NET (which includes `IHttpContextAccessor`).
And these will be added by either of the things shown above

**Reply by fseidl-bauradar** [reply]
Sounds a little bit fishy for me, replacing a deprecated Package with a deprecated Package with security holes in addition.

**Reply by linkdotnet** [reply]
That has nothing to do with the provided nuget packages you have shown.

There are old relicts from prior .net core 3 /.netstandard2.0. That is why they aren't supported anymore.
The things I did showcase here have nothing to do with that. Yes they have the same name - so that the API is stable, but that is it.

`IHttpContextAccessor` is very much alive and widely used.

Here the official documentation: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/target-aspnetcore?view=aspnetcore-8.0&tabs=visual-studio#use-the-aspnet-core-shared-framework

Be advised this is for net 8 - so very much alive and supported.

**Reply by fseidl-bauradar** [reply]
Thanks you for the Link, so I looking for a solution how to get ride of this Package since months now, seems that replacement.
